### PR TITLE
Support for URL Authentication

### PR DESCRIPTION
**Pull request by seanrucker**
With URL authentication the route will check for the token passed in as a query parameter and use it to authenticate before redirecting. A common use case for this is automatically authenticating a user from a link in a system generated email.

The name of the URL parameter is defined by the Auth.Config.tokenKey hook. 

Along with the regular set of sign in credentials, your token creation API end point should also accept the token itself.
e.g.
  if Auth.Config.tokenCreateUrl returns '/api/sign_in'
  and Auth.Config.tokenKey returns 'auth_token' 
  then a URL containing ?auth_token=fjlja8hfhf4
  will POST /api/sign_in { "auth_token": "fjlja8hfhf4" }

Until Ember supports query parameters the parameter must exist before the Ember route hash.
e.g.
  http://www.example.com/?auth_token=fjlja8hfhf4/#/posts/5

### CODE REVIEW

**Comment by seanrucker** [comment]
I couldn't get the tests to run, I was getting an error: "PhantomJS executable couldn't be auto detected.".

**Comment by heartsentwined** [comment]
Oh, my oversight, PhantomJS is also a dependency of Jasmine.

``` sh
# sudo apt-get install phantomjs
```

(or something similar for your system)

I'll add that to the readme.

Never mind if you can't get the tests to run; Travis will do that; and I'll also run them before every merge.

**Comment by seanrucker** [comment]
Due to Ember's lack of support for query string parameters there is a dependency on jQuery URL Parser: https://github.com/allmarkedup/jQuery-URL-Parser.

I'll keep an eye on https://github.com/emberjs/ember.js/issues/1773 and update the URL authentication support when Ember figures out how they want to handle the query string.

**Comment by seanrucker** [comment]
URL Authentication is done through another module, I mirrored the way remember-me works. I did move the monkey-patch logic from the remember-me module into the router. It seems a bit cleaner this way and gives us control over which order each module's logic is run. Not sure how you wanted to handle that, but let me know if you had another idea.

**Comment by heartsentwined** [comment]
This pull needs to be discussed in more detail though, it can get in the way of others' userland code.

First, the URL query string hack is known, and I bet there are already devs out there implementing their own logic to handle query string params. (viz. [ember-query](https://github.com/alexspeller/ember-query))

Second, this is performing an additional avoidable roundtrip to the server. Ideally, we should be providing a hook for the server to pre-load the session, not loading the app (_first roundtrip_) and then immediately sending another request (_second roundtrip_) for authentication.

I agree with your use case though, so I'd rather come up with a solution that avoids this roundtrip, much like `ember-data` allows us to pre-load models instead of "immediately fetching some data on app load".

**Comment by seanrucker** [comment]
Can you elaborate on "it can get in the way of others' userland code"?

Also, I'm not sure I understand how it is performing an avoidable roundtrip to the server. It works the same way as signing in with your email/pass or a remember-me token. It calls the same /api/sign_in endpoint passing the auth_token. There is only one call to sign in.

Presumably you could avoid the /api/sign_in call altogether since you already have the token; however you would need some other way to get the user id.

Thoughts?

Oops on the console.log :)

File: lib/routes/auth.coffee [hunk 1]
```diff
@@ -1,7 +1,20 @@
 Auth.Route = Em.Route.extend Em.Evented,
   redirect: ->
-    if !Auth.get 'authToken'
-      @trigger 'authAccess'
-      if Auth.Config.get 'authRedirect'
-        Auth.set 'prevRoute', @routeName
-        @transitionTo Auth.Config.get('signInRoute')
+    console.log 'redirect'
```
Comment (heartsentwined) [inline lib/routes/auth.coffee]: oh, and you left a `console.log` here :stuck_out_tongue:

**Comment by heartsentwined** [comment]
On potential interference with userland code: `ember-auth` should make as few assumptions and hacks on non auth-related stuff as possible. I am still digesting your code logic, so nothing concrete yet; but I would strive to avoid another #14, where hard-coded ajax settings in `ember-auth` interfered with userland code.

**Comment by heartsentwined** [comment]
On the additional roundtrip:

Your setup:
1. `GET /example.com?auth_token=abc/#/posts`. (_first roundtrip_)
2. Server returns response with HTML + assets.
3. Ember loads.
4. Ember-auth fires URL authentication. `POST /api/sign_in` (with auth token) (_second roundtrip_)
5. Server returns session details.
6. Ember redirects.

Alternative setup:
1. `GET /example.com?auth_token=abc/#/posts`. (_first roundtrip_)
2. Server already knows the auth token, so it can insert the session somewhere. In the JS file maybe? Or in the DOM somewhere? Or the headers? Either way, server returns response with HTML + assets (with session details).
3. Ember-auth loads the session.
4. Ember redirects.

**Comment by heartsentwined** [comment]
I am inclining towards letting the server set the session via headers. This way, we can let asset files remain static, hence cacheable and separately servable from CDNs. Response headers are expected to be different for each request already, so we could find a "best practice" to insert our required information, and retrieve it in `ember-auth`.

Some ideas: perhaps using the `X-` prefixed custom headers? Or cookies?

**Comment by seanrucker** [comment]
That assumes the server serving your HTML + assets is the same as your API server. This is not the case in my situation. My Ember application is served from an S3 bucket and my API server is a Rails application on Heroku. 

I think the alternative setup is a nice enhancement but I don't think we want to depend on the HTML and API server being the same.

**Comment by heartsentwined** [comment]
Valid point. I have also reviewed your code and it seems okay, so I take back on my "interfere with userland code".

Some more thoughts.
1. Which should take precedence, a local remember me cookie, or this?
2. Events. Standard `signIn-*` hooks, or special treatment? I am mostly concerned about errors / non-authenticated cases.
3. Does behavior differ / depend on whether the route in question is an `Auth.Route` or not?
4. Hence, on a failed sign in attempt (auth token expiry, for example), is it still treated as an `authAccess` failure? Or special treatment?

Thoughts on use cases:
1. Is this supposed to be a "share protected information with others" feature? e.g. Facebook has a "permalink" feature on its photos, such that I can send a photo normally - requiring my authentication to access - to others. In this case, allowing a full sign in of the original user would be overkill and potentially dangerous. This might be the server's concern though, e.g. generate a special auth token that is, e.g. valid only for the specified URL.
2. Or is this supposed to be a "sign in shortcut"? I'm not sure about the wisdom of allowing a URL end point that anybody could access. It acts, effectively, as an alternative password, that anybody who manages to get hold of this URL can access. But then, this would still be the server's concern, to make sure that this is a special auth token, probably one-off and short-expiring?
3. What if I want to _pause_ the sign in? Like, sending out a confirmation email, but require the user to enter one's password before proceeding? The URL could contain only a login identifier though, instead of a full auth-token.

The features discussed above do not block this PR; just something that came up in my mind when I saw your use case. Let me know if I'm reading your intended use case correctly.

How this interacts with other features needs clarification though.

**Comment by seanrucker** [comment]
USE CASE

This is a "sign in shortcut" feature. In my application users can send each other messages. When a user sends a message the application generates and delivers an email notification to the recipient. The email contains a link to the message detail page containing the auth_token. 

E.g. http://www.example.com/?auth_token=abc123/#/messages/5

This allows the user to view their message with one click, without having to enter their email/password.

You're right, if someone gets a hold of your token its essentially giving them the keys to your account. Therefore its a good idea for the tokens to expire after a certain amount of time, but this concern is left up to the server.

Devise supports this type of authentication out of the box with HTML apps. Passing the token in a URL parameter will automatically log the user in. 

PRECEDENCE

My thoughts were that the URL token would take precedence. This will allow you to switch accounts by changing the token in the URL.

Consider the scenario where you are using a shared computer. Someone on the shared computer logged into the application and a remember-me token is set. The user forgets to log off but walks away from the computer. A second user uses the same computer to log in to their email. They have a "new message" email notification in their inbox. They click the link: http://www.example.com/?auth_token=abc123/#/messages/5

If the remember-me cookie takes precedence over the URL authentication the previous user will remain logged in and the current user will not be authorized to access their message. 

If the URL authentication takes precedence the current user will be correctly logged in and the system works as expected.

EVENTS

Not sure what you were asking here. The way it is implemented, its using the same Auth.signIn function so everything should be handled the same way it is now. This is the same way remember-me is working.

AUTH.ROUTE

Not sure what you were asking here.

FAILED SIGN IN

Right now there is no special treatment. If the token is expired it would be treated as a failed sign-in, but I can see that would not be ideal in some cases.

**Comment by heartsentwined** [comment]
Thanks for your detailed answer. I'll merge it now, and tackle the remaining (edge?) use cases when issues come up. I'll also write up unit tests afterwards.

**Comment by heartsentwined** [comment]
This PR has been merged into version `3.1.0` release, along with (later added) unit tests and documentations.

**Comment by heartsentwined** [comment]
Note on BC Break: API introduced in this PR has been changed in the latest (`4.0.0`) release. URL authentication was re-written to support a more generic use case; you can now pass any param to the URL as expected by the server.

Two main changes:
- params are now scoped under (`Auth.Config.urlAuthenticationParamsKey`)
- (hence) it no longer depends on `Auth.Config.tokenKey` for the auth token

Relevant section from the upgrade guide in the [changelog](https://github.com/heartsentwined/ember-auth/blob/master/CHANGELOG.md):

> ### URL Authentication
> 
> Before:
> 
> ``` cofreescript
> Auth.Config.reopen
>   urlAuthentication: true
> ```
> 
> and a URL
> 
> ``` text
> http://www.example.com/?auth_token=fja8hfhf4/#/posts/5
> ```
> 
> After:
> 
> ``` coffeescript
> Auth.Config.reopen
>   urlAuthentication: true
>   urlAuthenticationParamsKey: 'auth' # or pick another name
> ```
> 
> and the corresponding URL
> 
> ``` text
> http://www.example.com/?auth[auth_token]=fja8hfhf4/#/posts/5
> ```

Full (updated) docs available at the Url Authentication section of the [readme](https://github.com/heartsentwined/ember-auth#url-authentication).

**Pull request merged**